### ISSUE TITLE
Linux Steam installation binary fails to run (segfault) [solved]

### ISSUE THREAD

**Issue by nyov** [endless-sky/endless-sky#5113]
**Describe the bug**

Current game installation from the Steam source segfaults for me.

Additionally, it does not package and provide the necessary libs that `endless-sky` is dynamically linked against. (I was only missing `libGLEW.so.1.10` myself, since I only had the newer libglew2.0 installed; but others like `libsdl` or such are also not a base installation requirement and potentially missing on the target system, or having a differing version.)

So before encountering the segfault, I also first had to install the missing libraries through my system package management to make it run (which wasn't a problem for me, but might be for people who don't know how to track the issue when the game simply fails to start from the steam launcher).
I don't mind not shipping (potentially outdated) local libs, but some notification about missing libs would be very helpful then. Steam just doesn't return anything, error-wise.

```
$ ./endless-sky 
Unsupported color conversion request
Unsupported color conversion request

Segmentation fault
```

**System (please complete the following information):**
 - OS: Debian GNU/Linux 10.4 (buster) / libc 2.28-10
 - Game Source: Steam
 - Version: 0.9.12 (according to credits.txt) / installation from May 27th

 - Game Source: Continuous Build
 - Version: 2020/05/23

**Additional context**
backtrace

```
$ gdb ./endless-sky
[...]
(gdb) run
Starting program: /[...]/Endless Sky/endless-sky 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff4ae9700 (LWP 2192)]
[New Thread 0x7ffff42e8700 (LWP 2193)]
[New Thread 0x7ffff3ae7700 (LWP 2194)]
[New Thread 0x7ffff32e6700 (LWP 2195)]
[New Thread 0x7ffff2ae5700 (LWP 2196)]
[New Thread 0x7ffff22e4700 (LWP 2197)]
[New Thread 0x7ffff1ae3700 (LWP 2198)]
[New Thread 0x7ffff12e2700 (LWP 2199)]
Unsupported color conversion request
Unsupported color conversion request
[Thread 0x7ffff3ae7700 (LWP 2194) exited]
[Thread 0x7ffff42e8700 (LWP 2193) exited]
[Thread 0x7ffff2ae5700 (LWP 2196) exited]

Thread 8 "endless-sky" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff1ae3700 (LWP 2198)]
0x00007ffff7123d57 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) backtrace
#0  0x00007ffff7123d57 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff7123eba in exit () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007ffff79f916c in ?? () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#3  0x00007ffff79f062d in jinit_color_deconverter () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#4  0x00007ffff79f7c8c in jinit_master_decompress () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#5  0x00007ffff79eceb5 in jpeg_start_decompress () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#6  0x00000000004d2d4b in (anonymous namespace)::ReadJPG(std::string const&, ImageBuffer&, int) ()
#7  0x00000000004d4b34 in ImageBuffer::Read(std::string const&, int) ()
#8  0x00000000004d6196 in ImageSet::Load() ()
#9  0x00000000005c7185 in SpriteQueue::operator()() ()
#10 0x000000000062c5f0 in execute_native_thread_routine ()
#11 0x00007ffff744ffa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#12 0x00007ffff71e34cf in clone () from /lib/x86_64-linux-gnu/libc.so.6
(gdb)

$ ldd endless-sky | grep jpeg
	libjpeg.so.8 => /usr/lib/x86_64-linux-gnu/libjpeg.so.8 (0x00007fbad0553000)
```

I've now tested the latest build from the github releases page and come across the same issue there.
(Except for different bt offsets there, in case those help:
```
(gdb) bt
#0  0x00007ffff7162d57 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff7162eba in exit () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007ffff7be716c in ?? () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#3  0x00007ffff7bde62d in jinit_color_deconverter () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#4  0x00007ffff7be5c8c in jinit_master_decompress () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#5  0x00007ffff7bdaeb5 in jpeg_start_decompress () from /usr/lib/x86_64-linux-gnu/libjpeg.so.8
#6  0x0000555555648ced in (anonymous namespace)::ReadJPG(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, ImageBuffer&, int) ()
#7  0x0000555555649f05 in ImageBuffer::Read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int) ()
#8  0x000055555564b842 in ImageSet::Load() ()
#9  0x000055555577ebef in SpriteQueue::operator()() ()
#10 0x00007ffff7542b2f in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007ffff7634fa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#12 0x00007ffff72224cf in clone () from /lib/x86_64-linux-gnu/libc.so.6
```
)

In case there was an issue with my local jpeg library (which is libjpeg.so.8.4.0 / version 8d1), I also tested loading an older version of libjpeg.so.8, but without any change.
I've also not found any issues with the jpeg art assets, though; so I'm stumped on what the issue is here.

So finally, I did a local debug build from git... and... it works.

**Comment by MCOfficer** [comment]
The bug might be a nightmare, but your report is a dream come true ;)

> So finally, I did a local debug build from git... and... it works.

Just to clarify: All precompiled binaries you tested fail, but all self-compiled builds work? Which versions did you try to compile yourself?

**Comment by nyov** [comment]
Thanks :) I know how unhelpful it would be to report a segfault without any info. And yes, that seems to be the case with the prebuilt binaries. I only compiled the latest git checkout.

I now noticed now that my local build actually built against `libjpeg 6`. So it's very much possibly an issue with my local libjpeg8.

```diff
$ diff -up \
	<(ldd continuous-prebuilt/endless-sky | sort | sed 's/(0x.*//') \
	<(ldd local/endless-sky | sort | sed 's/(0x.*//')
--- /dev/fd/63	2020-05-29 14:49:46.189616253 +0000
+++ /dev/fd/62	2020-05-29 14:49:46.193616251 +0000
@@ -11,11 +11,11 @@
 	libFLAC.so.8 => /usr/lib/x86_64-linux-gnu/libFLAC.so.8 
 	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 
 	libgcrypt.so.20 => /lib/x86_64-linux-gnu/libgcrypt.so.20 
-	libGLEW.so.2.0 => /usr/lib/x86_64-linux-gnu/libGLEW.so.2.0 
+	libGLEW.so.2.1 => /usr/lib/x86_64-linux-gnu/libGLEW.so.2.1 
 	libGL.so.1 => /usr/lib/x86_64-linux-gnu/libGL.so.1 
 	libgpg-error.so.0 => /lib/x86_64-linux-gnu/libgpg-error.so.0 
 	libICE.so.6 => /usr/lib/x86_64-linux-gnu/libICE.so.6 
-	libjpeg.so.8 => /usr/lib/x86_64-linux-gnu/libjpeg.so.8 
+	libjpeg.so.62 => /usr/lib/x86_64-linux-gnu/libjpeg.so.62 
 	liblz4.so.1 => /usr/lib/x86_64-linux-gnu/liblz4.so.1 
 	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 
 	libmad.so.0 => /usr/lib/x86_64-linux-gnu/libmad.so.0 
```

It seems debian doesn't really have libjpeg8 in stable and my version must have slipped in with some package from 'sid' (libjpeg8 and libjpeg9 are only in sid/unstable currently) at some point. And by now, libjpeg8-dev has been removed in favor of libjpeg9-dev in the unstable debian repo, so I can't really testbuild that.

So I've tested a hacky build against libjpeg**9**-dev headers [from debian unstable](https://packages.debian.org/sid/libjpeg9-dev), which errored out with
```
g++ -o build/release/ImageBuffer.o -c -std=c++11 -Wall -O3 source/ImageBuffer.cpp
source/ImageBuffer.cpp: In function ‘bool {anonymous}::ReadJPG(const string&, ImageBuffer&, int)’:
source/ImageBuffer.cpp:274:27: error: ‘JCS_EXT_BGRA’ was not declared in this scope
   cinfo.out_color_space = JCS_EXT_BGRA;
                           ^~~~~~~~~~~~
source/ImageBuffer.cpp:274:27: note: suggested alternative: ‘JCS_BG_RGB’
   cinfo.out_color_space = JCS_EXT_BGRA;
                           ^~~~~~~~~~~~
                           JCS_BG_RGB
scons: *** [build/release/ImageBuffer.o] Error 1
scons: building terminated because of errors.
```
It looks like a perfect match for the troublesome "Unsupported color conversion request" message that I see on the segfault: `In function ‘bool {anonymous}::ReadJPG`.
But I don't know why it would even compile for your build, if that's the reason, unless your libjpeg8 headers are older, or different in debian (which would be strange), or what. :/

**Comment by nyov** [comment]
On further investigation, `JCS_EXT_*` colorspaces are defined by [jpeg-turbo](https://github.com/ImageMagick/jpeg-turbo#libjpeg-v7-and-v8-apiabi-emulation).
It seems debian does not have any -turbo builds for libjpeg8 (or libjpeg9) at present.

I suppose packaging your libjpeg.so with the game may be the only way to solve the problem for me, currently (other than using local builds).

**Comment by MCOfficer** [comment]
that's correct, i can't speak for the Releases (built by MZ), but the Continuous builds link against the packages mentioned in [BuildInstructions](https://github.com/endless-sky/endless-sky/wiki/BuildInstructions), with the versions being those of the oldest Ubuntu LTS (currently 16.04).

You reckon we can do something about this? @tehhowch

**Comment by RadhiFadlillah** [comment]
Hi guys, I've experienced this issue as well in Manjaro Linux. It works when I install from AUR, however it failed when run from Steam :

```
$ ./endless-sky
./endless-sky: error while loading shared libraries: libGLEW.so.1.10: cannot open shared object file: No such file or directory
`````

**Comment by MCOfficer** [comment]
Unless you forgot to post the important part, this does not look like the same issue. In that case, please open a new one.

**Comment by RadhiFadlillah** [comment]
@MCOfficer ah yeah, sorry.

I was planning to make a new issue, however after reading this bug report, I found out our initial problem is identical i.e. Endless Sky in Steam can't be started because it's missing some of linked dependencies.

I've opened a new issue in #5135.

**Comment by nyov** [comment]
Well, the core problem is certainly the same: the program fails to ship the libraries it depends on, to run.

**Comment by tehhowch** [comment]
@nyov 
this part:
> So before encountering the segfault, I also first had to install the missing libraries through my system package management to make it run
Makes me think you were attempting to run the binary that was compiled for the Steam runtime _outside of the Steam runtime_. That's not supported. If you want to run the Steam-provided binary without launching the Steam client, you still need to tell the game where to find all of the libraries.

from https://github.com/endless-sky/endless-sky/issues/5135#issuecomment-855309216 :
```
Endless Sky $ ./endless-sky -v
./endless-sky: error while loading shared libraries: libpng12.so.0: cannot open shared object file: No such file or directory
Endless Sky $ ../../../ubuntu12_32/steam-runtime/run.sh ./endless-sky -v

Endless Sky ver. 0.9.13
License GPLv3+: GNU GPL version 3 or later: <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Compiled against SDL v2.0.14
Using SDL v2.0.14

Endless Sky $ pwd
~/.local/share/Steam/steamapps/common/Endless Sky
```
Basically, you can find a runner script in the steam runtime directory that will let you launch the Steam-provided binary without first launching the Steam UI.

Local builds work because you're invoking them from the same environment in which they were built.

Closing this since it seems to be only an issue when using the Steam binary incorrectly, rather than a scenario that is expected to be supported.
